tool/tctl/common/loadtest_command.go

```diff
@@ -0,0 +1,148 @@
+
+	count       int
+	interval    time.Duration
+	ttl         time.Duration
+	concurrency int
+}
+
+// Initialize allows LoadtestCommand to plug itself into the CLI parser
+func (c *LoadtestCommand) Initialize(app *kingpin.Application, config *servicecfg.Config) {
+	c.config = config
+	loadtest := app.Command("loadtest", "Tools for generating artificial load").Hidden()
+
+	c.nodeHeartbeats = loadtest.Command("node-heartbeats", "Generate artifical node heartbeats").Hidden()
+	c.nodeHeartbeats.Flag("count", "Number of unique nodes to heartbeat").Default("10000").IntVar(&c.count)
+	c.nodeHeartbeats.Flag("interval", "Node heartbeat interval").Default("1m").DurationVar(&c.interval)
+	c.nodeHeartbeats.Flag("ttl", "TTL of heartbeated nodes").Default("10m").DurationVar(&c.ttl)
+	c.nodeHeartbeats.Flag("concurrency", "Max concurrent requests").Default(
+		strconv.Itoa(runtime.NumCPU() * 16),
+	).IntVar(&c.concurrency)
+}
+
+// TryRun takes the CLI command as an argument (like "loadtest node-heartbeats") and executes it.
+func (c *LoadtestCommand) TryRun(ctx context.Context, cmd string, client auth.ClientI) (match bool, err error) {
+	switch cmd {
+	case c.nodeHeartbeats.FullCommand():
+		err = c.NodeHeartbeats(ctx, client)
+	default:
+		return false, nil
+	}
+	return true, trace.Wrap(err)
+}
+
+func (c *LoadtestCommand) NodeHeartbeats(ctx context.Context, client auth.ClientI) error {
+	infof := func(format string, args ...any) {
+		fmt.Fprintf(os.Stderr, "[i] "+format+"\n", args...)
+	}
+	warnf := func(format string, args ...any) {
+		fmt.Fprintf(os.Stderr, "[!] "+format+"\n", args...)
+	}
+
+	infof("Setting up node hb load generation. count=%d, interval=%s, ttl=%s, concurrency=%d",
+		c.count,
+		c.interval,
+		c.ttl,
+		c.concurrency,
+	)
+
+	node_ids := make([]string, 0, c.count)
+
+	for i := 0; i < c.count; i++ {
+		node_ids = append(node_ids, uuid.New().String())
+	}
+
+	workch := make(chan string, c.count*2)
+
+	errct := new(atomic.Uint64)
+	errch := make(chan error, 1)
+
+	for i := 0; i < c.concurrency; i++ {
+		go func() {
+			for id := range workch {
+				node, err := types.NewServer(id, types.KindNode, types.ServerSpecV2{})
+				if err != nil {
+					panic(err)
+				}
+				node.SetExpiry(time.Now().Add(c.ttl))
+
+				_, err = client.UpsertNode(ctx, node)
+				if err != nil {
+					select {
+					case errch <- err:
+					default:
+					}
+					errct.Add(1)
+				}
+			}
+		}()
+	}
+
+	ticker := time.NewTicker(c.interval)
+	var generation uint64
+	for {
+		if backlog := len(workch); backlog != 0 {
+			warnf("Backlog in heartbeat emission. size=%d", backlog)
+		}
+
+		select {
+		case err := <-errch:
+			warnf("Error during last round of heartbeats: %v", err)
+		default:
+		}
+
+		for _, id := range node_ids {
+			select {
+			case workch <- id:
+			default:
+				panic("too much backlog!")
+			}
+		}
+
+		generation++
+
+		infof("Queued heartbeat batch for emission. generation=%d, errors=%d", generation, errct.Load())
+
+		<-ticker.C
```

Comment:
It would be nice to cleanly stop this on a ^C, maybe with a `signal.NotifyContext`.

Reply:
+1